astropy/io/fits/tests/test_fsspec.py

```diff
@@ -0,0 +1,128 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Can `astropy.io.fits.open` access (remote) data using the fsspec package?
+"""
+from astropy.io import fits
+from astropy.nddata import Cutout2D
+from astropy.utils.compat.optional_deps import HAS_FSSPEC, HAS_S3FS  # noqa
+from astropy.utils.data import get_pkg_data_filename, conf
+
+import numpy as np
+from numpy.testing import assert_array_equal, assert_allclose
+import pytest
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_local():
+    """Can we use fsspec to read a local file?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    hdulist_classic = fits.open(fn)
+    hdulist_fsspec = fits.open(fn, use_fsspec=True)
+    assert_array_equal(hdulist_classic[2].data, hdulist_fsspec[2].data)
+    assert_array_equal(hdulist_classic[2].section[3:5], hdulist_fsspec[2].section[3:5])
+    hdulist_classic.close()
+    hdulist_fsspec.close()
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_local_write(tmpdir):
+    """Can we write to a local file that was opened using fsspec?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    fn_tmp = tmpdir / "tmp.fits"
+    with fits.open(fn, use_fsspec=True) as hdul:
+        # writing to a section is never allowed
+        with pytest.raises(TypeError):
+            hdul[1].section[0, 0] = -999
+        # however writing to .data should work
+        hdul[1].data[2, 3] = -999
+        assert hdul[1].data[2,3] == -999
+        hdul.writeto(fn_tmp)
+
+    # Is the new value present when we re-open the file?
+    with fits.open(fn_tmp) as hdul:
+        assert hdul[1].data[2,3] == -999
+
+    # Does fsspec support `mode="update"` for local files?
+    with fits.open(str(fn_tmp), use_fsspec=True, mode="update") as hdul:
+        hdul[1].data[2,3] = 42
+    with fits.open(fn_tmp) as hdul:
+        assert hdul[1].data[2,3] == 42
+
+
+@pytest.mark.remote_data
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_http():
+    """Can we use fsspec to open a remote FITS file via http?"""
```

Comment:
note: I very much like the one-liner test docs, I bet these will be super helpful down the line for maintenance.